### ISSUE TITLE
Integration with signal-hook

### ISSUE THREAD

**Issue by vorner** [alexcrichton/tokio-signal#45]
Hello

One of the goals when I was writing signal-hook was to be able to share the signals between different libraries in some sane way. Tokio-signal is one of the more famous libraries that do signals. Do you think it would make sense to use signal-hook as a backend instead of registering the signals directly?

I'm willing to send the pull request, I'm just asking if there is any chance it could get accepted or if there's some kind of philosophical (or different) reason such concept wouldn't fly.

**Comment by vorner** [comment]
Hello

1. The signal chaining works, but all the libraries that use it must support it (and many don't, tokio-signal being the good exception). Besides, the chaining looks a bit complex/messy. And there's no reasonable way to *remove* signals from the chain if someone wanted.

2. With the architecture, I actually got inspired somewhat by `tokio-signal`. The chaining code is mostly stolen from there. The big difference is that there's a map of registered signals inside and `signal-hook` iterates through all the callbacks registered for that one signal (as well as chaining to the previously registered signal handler). This allows for dynamically adding and removing hooks to a single signal without influencing other signal callbacks.  

   The hard way is updating the map when there's a need to add or remove a callback. Because one can't use locks inside a signal handler, an update is done by creating a brand new copy of the map, making the updates and then swapping it. Then the updater busy-waits until all previous signal handlers terminate before throwing the old map away (to make sure the signal handler doesn't have to free it). So, if there's some long-running callback or there are a lot of callbacks (so copying of the map takes time), the update is potentially expensive operation.

3. I don't really have a definite answer to that. I don't know if `tokio-signal` supports the new `tokio` or just `tokio-core` (it's some time I used the crate or touched the code). But it might make some sense to either provide tokio-signal as API-compat over tokio-signal and make it effectively „hollow“ or to drop the tokio-support from signal-hook and instead of adding the code needed by a flag, doing it by adding a crate. But I think time would answer that.

I also noticed some PR to get rid of the complex code for re-broadcasting between different reactors which seems to also be a bit complex (I haven't followed it *closely*, but I know there are some issues around there). I believe the fact that `signal-hook` allows both adding and removing would make this problem go away by effectively not sharing the signal callback between different streams ‒ each one could create its own callback, register it on creation and unregister on drop. That would make the signal handler itself broadcast it ‒ that seems simpler to implement, but has potentially worse performance, as there are more self-pipes involved.

**Comment by ipetkov** [comment]
Thanks for the follow up info @vorner!

1. I don't anticipate `tokio-signal` needing to deregister signal handlers, but I suppose it doesn't hurt to have the option to do so!
2. It may be worth while to figure out a way such that `tokio-signal` wouldn't need to constantly (re)register signals (maybe we can install a single handler that fires on all signals, lazily register handlers for requested signals without deregistering). One reason for this is currently `tokio-process` works by creating a new signal instance for every new spawned process. Obviously we can figure out ways to optimize this if needed, but some food for thought.
3. We definitely support both `tokio` and `tokio-core` today (really any executor as long as there is a running tokio `Reactor` somewhere). The current design creates a Driver for each spawned Signal instance which generates a lot of unnecessary wake ups. I had tried to put this in a single global task (which would solve the pipe contention issues and needless wakeups), but it turned out to be a dead end since `tokio` can't guarantee it won't leak futures when dropped (i.e. we can't rely on Drop to tell us we need to make a new global task).

All and all, I'd be happy to review a PR for this! I can see some issues that would need to be ironed out/tested further but we can revisit as we get to them.

**Comment by ipetkov** [comment]
AFAIK there's no way to identify/compare executors so I'm not sure how to identify if the current executor has a running driver or not

**Comment by ipetkov** [comment]
@vorner With #50 merged into master, are there any other immediate `signal-hook` integrations that should be done before moving the repo/issue tracker over?

**Comment by vorner** [comment]
I want to go through the code and look if there's a space to simplify/use some of the features of signal-hook there eventually, but I don't have anything in the pipeline right now. I think this one can be closed for now ‒ I believe anything more would generally fall in the category of refactoring.
